// nuxt-plotly/dist/runtime/components/nuxt-plotly.mjs
import {
  h,
  defineComponent,
  reactive,
  onBeforeUnmount,
  onMounted,
  watch
} from "vue";
import { v4 as uuidv4 } from "uuid";
import Plotly from "plotly.js-dist-min";
export default defineComponent({
  props: {
    data: { type: Array, required: true },
    config: {
      type: Object,
      required: false,
      default: void 0
    },
    layout: {
      type: Object,
      required: false,
      default: void 0
    }
  },
  emits: ["on-ready"],
  setup(props, ctx) {
    const datas = reactive({
      plotlyId: `plotly-${uuidv4()}`,
      resizeObserver: {},
      timeOutFunctionId: {},
      plotlyHTMLElement: {}
    });
    const setGraph = async () => {
      datas.plotlyHTMLElement = await Plotly.newPlot(
        datas.plotlyId,
        props.data,
        props.layout,
        props.config
      );
      ctx.emit("on-ready", datas.plotlyHTMLElement);
    };
    const setResizeObserver = () => {
      datas.resizeObserver = new ResizeObserver(() => {
        clearTimeout(datas.timeOutFunctionId);
        datas.timeOutFunctionId = setTimeout(() => {
          setGraph();
        }, 100);
      });
      const plotlyElm = document.getElementById(datas.plotlyId);
      if (plotlyElm) {
        datas.resizeObserver.observe(plotlyElm);
      }
    };
    onMounted(() => {
      setGraph();
      setResizeObserver();
    });
    watch(
      () => [props.data, props.layout, props.config],
      () => {
        setGraph();
      },
      { deep: true }
    );
    onBeforeUnmount(() => {
      datas.resizeObserver.disconnect();
    });
    return () => h("div", { id: datas.plotlyId });
  }
});
